src/tree/gpu_hist/histogram.cu

```diff
@@ -0,0 +1,155 @@
+/*!
+ * Copyright 2020 by XGBoost Contributors
+ */
+#include <thrust/reduce.h>
+#include <algorithm>
+#include <ctgmath>
+
+#include "xgboost/base.h"
+#include "row_partitioner.cuh"
+
+#include "histogram.cuh"
+
+#include "../../data/ellpack_page.cuh"
+#include "../../common/device_helpers.cuh"
+
+namespace xgboost {
+namespace tree {
+// Following 2 functions are slightly modifed version of fbcuda.
+
+/* \brief Constructs a rounding factor used to truncate elements in a sum such that the
+   sum of the truncated elements is the same no matter what the order of the sum is.
+
+ * Algorithm 5: Reproducible Sequential Sum in 'Fast Reproducible Floating-Point
+ * Summation' by Demmel and Nguyen
+
+ * In algorithm 5 the bound is calculated as $max(|v_i|) * n$.  Here we use the bound
+ *
+ * \begin{equation}
+ *   max( fl(\sum^{V}_{v_i>0}{v_i}), fl(\sum^{V}_{v_i<0}|v_i|) )
+ * \end{equation}
+ *
+ * to avoid outliers, as the full reduction is reproducible on GPU with reduction tree.
+ */
+DEV_INLINE __host__ float CreateRoundingFactor(float max_abs, int n) {
+  float delta = max_abs / (1.0f - 2 * n * FLT_EPSILON);
+
+  // Calculate ceil(log_2(delta)).
+  // frexpf() calculates exp and returns `x` such that
+  // delta = x * 2^exp, where `x` in (-1.0, -0.5] U [0.5, 1).
+  // Because |x| < 1, exp is exactly ceil(log_2(delta)).
+  int exp;
+  frexp(delta, &exp);
+
+  // return M = 2 ^ ceil(log_2(delta))
+  return ldexp(1.0f, exp);
+}
+
+GradientPair CreateRoundingFactor(common::Span<GradientPair> gpair) {
+  auto positive = []__device__(float v) {
+    return v > 0 ? v : 0;
+  };  // NOLINT
+  auto negative = []__device__(float v) {
+    return v < 0 ? abs(v) : 0;
+  };  // NOLINT
+
+  dh::XGBCachingDeviceAllocator<char> alloc;
+  auto positive_sum = thrust::reduce(
```

Comment:
Not sure if it matters, but it is possible to fuse these two reductions into a single reduction e.g. apply transform iterator to gradient pair inputs that returns 2 GradientPair (pos/neg) and define an operator to sum them. Alternatively just use the absolute sum, although this is slightly less tight and we could lose up to one bit of precision.

Reply:
@RAMitchell I tried using `cub::DeviceReudce` with 2 cuda streams for this, the end result doesn't seem to be any different.  As for abs sum, it failed a Windows test for GPU ranking when using float.